subsys/power/reboot.c

```diff
@@ -33,3 +36,46 @@ void sys_reboot(int type)
 		k_cpu_idle();
 	}
 }
+
+#ifdef CONFIG_ARDUINO_LIKE_UPGRADE
+
+void sys_reboot_to_upgrade(int delay_ms)
+{
+	k_sleep(Z_TIMEOUT_MS(delay_ms));
+
+	(void)irq_lock();
+#ifdef CONFIG_SYS_CLOCK_EXISTS
+	sys_clock_disable();
+#endif
+
+#ifdef CONFIG_SOC_SERIES_SAMD21
+	while (!NVMCTRL->INTFLAG.reg & NVMCTRL_INTFLAG_READY) {
```

Comment:
Instead of erasing the flash, you can put the DOUBLE_TAP_MAGIC value at the top of RAM.